catalog/from-k8s/resource_test.go

```diff
@@ -1770,3 +1770,297 @@ func testService(name string) *apiv1.Service {
+	require.Equal("b", actual[1].Service.Address)
+}
+
+// Test node port works with prefix
+func TestServiceResource_nodePortPrefix(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+	client := fake.NewSimpleClientset()
+	syncer := &TestSyncer{}
+
+	// Start the controller
+	closer := controller.TestControllerRun(&ServiceResource{
+		Log:             hclog.Default(),
+		Client:          client,
+		Syncer:          syncer,
+		NodePortSync:    ExternalOnly,
+		K8sConsulPrefix: "prefix",
+	})
+	defer closer()
+
+	node1 := "ip-10-11-12-13.ec2.internal"
+	node2 := "ip-10-11-12-14.ec2.internal"
+	// Insert the nodes
+	_, err := client.CoreV1().Nodes().Create(&apiv1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: node1,
+		},
+
+		Status: apiv1.NodeStatus{
+			Addresses: []apiv1.NodeAddress{
+				apiv1.NodeAddress{Type: apiv1.NodeExternalIP, Address: "1.2.3.4"},
+				apiv1.NodeAddress{Type: apiv1.NodeInternalIP, Address: "4.5.6.7"},
+			},
+		},
+	})
+	require.NoError(err)
+
+	_, err = client.CoreV1().Nodes().Create(&apiv1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: node2,
+		},
+
+		Status: apiv1.NodeStatus{
+			Addresses: []apiv1.NodeAddress{
+				apiv1.NodeAddress{Type: apiv1.NodeExternalIP, Address: "2.3.4.5"},
+				apiv1.NodeAddress{Type: apiv1.NodeInternalIP, Address: "3.4.5.6"},
+			},
+		},
+	})
+	require.NoError(err)
+	time.Sleep(200 * time.Millisecond)
+
+	// Insert the endpoints
+	_, err = client.CoreV1().Endpoints(metav1.NamespaceDefault).Create(&apiv1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+		},
+
+		Subsets: []apiv1.EndpointSubset{
+			apiv1.EndpointSubset{
+				Addresses: []apiv1.EndpointAddress{
+					apiv1.EndpointAddress{NodeName: &node1, IP: "1.2.3.4"},
+				},
+				Ports: []apiv1.EndpointPort{
+					apiv1.EndpointPort{Name: "http", Port: 8080},
+					apiv1.EndpointPort{Name: "rpc", Port: 2000},
+				},
+			},
+
+			apiv1.EndpointSubset{
+				Addresses: []apiv1.EndpointAddress{
+					apiv1.EndpointAddress{NodeName: &node2, IP: "2.3.4.5"},
+				},
+				Ports: []apiv1.EndpointPort{
+					apiv1.EndpointPort{Name: "http", Port: 8080},
+					apiv1.EndpointPort{Name: "rpc", Port: 2000},
+				},
+			},
+		},
+	})
+	require.NoError(err)
+
+	// Wait a bit
+	time.Sleep(300 * time.Millisecond)
+
+	// Insert the service
+	_, err = client.CoreV1().Services(metav1.NamespaceDefault).Create(&apiv1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+		},
+
+		Spec: apiv1.ServiceSpec{
+			Type: apiv1.ServiceTypeNodePort,
+			Ports: []apiv1.ServicePort{
+				apiv1.ServicePort{Name: "http", Port: 80, TargetPort: intstr.FromInt(8080), NodePort: 30000},
+				apiv1.ServicePort{Name: "rpc", Port: 8500, TargetPort: intstr.FromInt(2000), NodePort: 30001},
+			},
+		},
+	})
+	require.NoError(err)
+
+	// Wait a bit
+	time.Sleep(300 * time.Millisecond)
+
+	// Verify what we got
+	syncer.Lock()
+	defer syncer.Unlock()
+	actual := syncer.Registrations
+	require.Len(actual, 2)
```

Comment:
To cut down on the amount of checks one thing you could do is declare a struct with the output you expect and let `require.Equal` do all the comparing:
```go
expected := []*api.CatalogRegistration{
	&api.CatalogRegistration{
		Service: &api.AgentService{
			Service: "prefixfoo",
			Address: "1.2.3.4",
			Port:    30000},
		Node: "k8s-sync",
	},
	&api.CatalogRegistration{
		Service: &api.AgentService{
			Service: "prefixfoo",
			Address: "2.3.4.5",
			Port:    30000},
		Node: "k8s-sync",
	},
}
require.Equal(t, expected, actual)
```

I find that it can help clarify the values and structure of what you're expecting.

Also, if you declare a struct like this in the beginning of the test, you can substitute hardcoded values like the addresses in lines 1935 and 1945 for `expected[0].Service.Address` and `expected[1].Service.Address`

Reply:
Actually, nevermind, this wouldn't really work here because not all the fields will match. (like ID)